File: 2025-04-07/forkjoin/src/test/java/MaxForkJoinTest.java
import org.junit.Test;

import java.util.Random;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class MaxForkJoinTest {

    private double sequentialMax(int[] arr) {
        double max = Double.NEGATIVE_INFINITY;
        for (int i=0; i < arr.length; i++) {
            double sinValue = Math.sin(arr[i]);
            if (sinValue > max) {
                max = sinValue;
            }
        }
        return max;
    }

    @Test
    public void testMax() throws Exception {
        Random r = new Random(90125);
        int[] arr = new int[10_000_000];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = r.nextInt();
        }

        // Do the sequential test a number of times to warm it up, measure on
        // last
        long timeBefore = 0;
        int numWarmups = 3;
        long seqTime = 0;
        double seqMax = 0;
        for (int i=0; i < numWarmups; i++) {
            timeBefore = System.currentTimeMillis();
            seqMax = sequentialMax(arr);
            seqTime = System.currentTimeMillis() - timeBefore;
        }

        double parallelMax = 0;
        long parallelTime = 0;
        // Do the parallel test a number of times to warm it up, measure on last
        // int[] seqCutoffs = {1, 2, 10, 100, 1000, 10000};
        int[] seqCutoffs = {
            1, 2, 3, 4, 10, 100, 1000, 10000, 100_000, 1_000_000, 5_000_000
        };
        System.out.println("cutoff / seq time / parallel time / speedup");
        for (int seqCutoff : seqCutoffs) {
            // Do the parallel test a number of times to warm it up, measure on
            // last
            for (int i = 0; i < numWarmups; i++) {
                timeBefore = System.currentTimeMillis();
                parallelMax = MaxForkJoin.max(arr, seqCutoff);
                parallelTime = System.currentTimeMillis() - timeBefore;
            }
            assertEquals(seqMax, parallelMax, 1e-9);
            double speedup = ((double) seqTime) / parallelTime;
            System.out.printf("%10d %5d %5d %7.3f\n", seqCutoff, seqTime,
                              parallelTime, speedup);
        }
    }
}

File: 2025-04-07/forkjoin/src/test/kotlin/MaxForkJoinTest.kt
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.random.Random
import kotlin.time.measureTimedValue


class MaxForkJoinTest {

    fun sequentialMax(arr: List<Int>): Int {
        var ans = Int.MIN_VALUE
        for (i in 0..<arr.count()) {
            if (isPrime(i) && i > ans) {
                ans = i
            }
        }
        return ans
    }

    @Test
    fun testMax() {
        var r = Random(90125)
        val arr = mutableListOf<Int>()
        for (i in 0..<100_000) {
            arr.add(r.nextInt(10000))
        }

        // Do the sequential test a number of times to warm it up, then do it
        // one more time and measure on last
        val numWarmups = 3
        // total each time; display it to make sure it doesn't get
        // optimized out
        var total = 0.0
        for (i in 0..<numWarmups) {
            total += sequentialMax(arr)
        }
        val (seqMax, seqTime) = measureTimedValue {
            sequentialMax(arr)
        }
        // Print the warmup total just to make sure that the code
        // doesn't get optimized out
        println("Warmup sequential max total = ${total}")
        println("Largest prime in list = ${seqMax}")
        println("Measured sequential time = ${seqTime}")

        val numProcessors = Runtime.getRuntime().availableProcessors()
        println("Number of processors = ${numProcessors}")
        println("-----------------------------")
        val seqCutoffs = listOf(
            1, 2, 3, 4, 10, 100, 1000, 10_000, 50_000, 100_000
        )
        println("cutoff / seq time / parallel time / speedup")

        for (seqCutoff in seqCutoffs) {
            // Do the parallel test a number of times to warm it up, measure on
            // last
            total = 0.0
            for (i in 0..<numWarmups) {
                total += sumMultiThreaded(arr, seqCutoff)
            }
            val (parallelMax, parallelTime) = measureTimedValue {
                maxMultiThreaded(arr, seqCutoff)
            }

            val speedup = seqTime/parallelTime
            println(String.format("%10d %s %s %7.3f", seqCutoff, seqTime,
                                  parallelTime, speedup))
            assertEquals(seqMax, parallelMax)
        }
    }
}
